### ISSUE TITLE
Over-eager -Wunused-but-set-variable for libc++ containers used only for side effect

### ISSUE THREAD

**Issue by stbergmann** [llvm/llvm-project#215545]
Since <https://github.com/llvm/llvm-project/commit/97f332f8d1274f74e02b2445a7c7317a4b458cb2> "[libc++] Diagnose variables of container types" on LLVM 24 trunk:

`$ cat test.cc`

```c++
#include <vector>
struct S { ~S(); };
void f(std::vector<S> & v) {
  std::vector<S> t;
  { // e.g., take a lock
    t = v;
  }
}
```

```
$ clang++ -Wunused-but-set-variable -fsyntax-only test.cc
test.cc:4:18: warning: variable 't' set but not used [-Wunused-but-set-variable]
    4 |   std::vector<S> t;
      |                  ^
1 warning generated.
```

where the sole reason for `t` is to run the `S` destructors.  (Real code would swap `t` and `v`, or something like that; just wanted to give a fairly minimal reproducer.)

**Comment by philnik777** [comment]
Do you have an actual example? We are aware that there could be false-positives, but believe that they should be _very_ rare. AFAICT swapping the containers would silence the warning, so that example doesn't apply.

**Comment by stbergmann** [comment]
See <https://gerrit.collaboraoffice.com/c/online/+/8989/1> "Hack to silence libc++ -Werror,-Wunused-but-set-variable", the relevant code is

```c++
typedef std::unordered_set< Reference<XInterface > > HashSet_Ref;
```

and

```c++
    // dispose
#if defined _LIBCPP_VERSION && _LIBCPP_VERSION == 240000
    [[maybe_unused]] // hack to silence libc++ -Werror,-Wunused-but-set-variable
#endif
    HashSet_Ref aImplMap;
    {
        MutexGuard aGuard( m_aMutex );
        // erase all members
        m_ServiceMap = HashMultimap_OWString_Interface();
        aImplMap = m_ImplementationMap;
        m_ImplementationMap = HashSet_Ref();
        m_ImplementationNameMap = HashMap_OWString_Interface();
        m_SetLoadedFactories= HashSet_Ref();
    }
```

so indeed using copy assignment rather than swapping.

**Comment by stbergmann** [comment]
...and swapping would indeed suppress any false warning here, so I'll go with that.  (And thus feel free to close this issue again if you're confident that any remaining false positives should be considered acceptable.)

**Comment by frederick-vs-ja** [comment]
I guess you can also rewrite the code snippet like the following.

```C++
    HashSet_Ref aImplMap = [this] {
        MutexGuard aGuard( m_aMutex );
        // erase all members
        m_ServiceMap = HashMultimap_OWString_Interface();
        HashSet_Ref retMap = std::exchange(m_ImplementationMap, HashSet_Ref());
        m_ImplementationNameMap = HashMap_OWString_Interface();
        m_SetLoadedFactories= HashSet_Ref();
        return retMap;
    }();
```

It seems to me that you wanted to clear `m_ImplementationMap` earlier but make the elements destroyed later. Not sure whether this was really intended.

**Comment by ldionne** [comment]
I think I agree that this is not worth a change to libc++ (per the discussion above), so I'll close this.

However, @philnik777 let's stay very alert about reports like this one. If we get too many, or if we get some where there's a genuine false positive, we'll have to make the attribute opt-in or find another solution. We don't want to start triggering a bunch of unhelpful diagnostics for users, that would just be enraging.

That's not the case here, but let's stay alert.

**Comment by philnik777** [comment]
> I think I agree that this is not worth a change to libc++ (per the discussion above), so I'll close this.
> 
> However, [@philnik777](https://github.com/philnik777) let's stay very alert about reports like this one. If we get too many, or if we get some where there's a genuine false positive, we'll have to make the attribute opt-in or find another solution. We don't want to start triggering a bunch of unhelpful diagnostics for users, that would just be enraging.
> 
> That's not the case here, but let's stay alert.

Yes, but I also think that there have been severe enough bugs found by this in LLVM alone (a potential use-after-free IIRC) that IMO we should accept a non-zero false-positive rate. Some experience with this on large code bases (CC @nico @vonosmas @yuxuanchen1997) would be nice to get some sense of the false-positive rate.

**Comment by yuxuanchen1997** [comment]
We haven't tried to build code with this patch yet. The next version we release internally by the end of the year would be libc++ 23 at best. We can try to incorporate this patch by then to see if there are too many false positives.
